File: src-tauri/plugins/tauri-plugin-levelplay/ios/Sources/tauri-plugin-levelplay/LevelPlayPlugin.swift
// iOS bridge for Unity LevelPlay (ironSource). Mirror of the Android
// plugin at ../../android/src/main/java/com/plugin/levelplay/LevelPlayPlugin.kt.
//
// COPPA / Families-Policy ordering rules (same as Android):
//
//   1. Every `LevelPlay.setMetaDataWithKey(...)` call MUST happen BEFORE
//      `LevelPlay.initWith(...)`. The init call propagates the flags to
//      each adapter's SDK during startup; setting metadata afterwards
//      is a silent no-op for adapters that have already initialised.
//
//   2. `LPMRewardedAd` / `LPMInterstitialAd` instances have their
//      delegate attached BEFORE we call `loadAd()` so the first load
//      attempt never races past the listener registration.
//
// LevelPlay 9.x uses per-instance ad objects keyed by dashboard-issued
// ad unit IDs. Each show method stores the active `Invoke` in a
// pending slot and resolves it exactly once from either `didCloseAd`
// or `didFailToDisplayAd`, so a concurrent second show can't
// cross-resolve an earlier caller's Promise.
//
// Apple Kids Category compliance:
//   • We never call ATTrackingManager.requestTrackingAuthorization —
//     apps in the Kids Category are forbidden from showing the ATT
//     prompt and from collecting IDFA. The `is_deviceid_optout=true`
//     metadata flag tells LevelPlay to use a per-install random UUID
//     instead of IDFA, mirroring the Android behaviour we verified
//     in production logcat.

import SwiftRs
import Tauri
import UIKit
import WebKit
// The IronSource xcframework's Swift module name is `IronSource`
// (per Unity's official LevelPlay-Swift integration guide), even
// though our SPM binary target is named `IronSourceSDK`. Linking
// comes from the binary-target dep in Package.swift; the import
// here resolves to the framework's modulemap.
//
// Exposes: `LevelPlay`, `LPMRewardedAd`, `LPMInterstitialAd`,
// `LPMInitRequestBuilder`, `LPMRewardedAdDelegate`,
// `LPMInterstitialAdDelegate`, `LPMAdInfo`, `LPMAdError`, `LPMReward`,
// `LPMConfiguration`.
import IronSource

// JS → native payload for the initialize command. Shape matches
// `models::InitRequest` on the Rust side; Tauri auto-deserializes.
class InitArgs: Decodable {
    let appKey: String
    var rewardedAdUnitId: String = ""
    var interstitialAdUnitId: String = ""
    var isChildDirected: Bool = true
    var admobTfcd: Bool = true
    var admobTfua: Bool = true
    var deviceIdOptOut: Bool = true
    var metaMixedAudience: Bool = false
    var enableTestSuite: Bool = false
}

class LevelPlayPlugin: Plugin,
    LPMRewardedAdDelegate,
    LPMInterstitialAdDelegate
{
    private var initialized = false
    private var rewardedAd: LPMRewardedAd?
    private var interstitialAd: LPMInterstitialAd?
    // The 9.x ad objects don't expose the adUnitId they were created with as
    // a public property, so we mirror it here to dispatch shared delegate
    // callbacks (didFailToLoadAd fires on both formats with the same
    // signature) to the correct format.
    private var rewardedAdUnitId: String?
    private var interstitialAdUnitId: String?
    private var pendingRewarded: Invoke?
    private var pendingInterstitial: Invoke?
    private var lastRewardGranted = false

    // Resolves the active webview's containing UIViewController. Tauri's
    // Plugin base exposes `manager` as a non-optional `PluginManager`,
    // but its `viewController` accessor may be nil in edge cases (no
    // scene yet, or the plugin instantiated before the webview is
    // attached). Fall back to UIApplication's key window root VC.
    @MainActor
    private var topViewController: UIViewController? {
        if let vc = self.manager.viewController { return vc }
        return UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first(where: \.isKeyWindow)?
            .rootViewController
    }

    // Emit a per-format readiness event to the webview. The JS provider
    // listens for "rewarded-state" / "interstitial-state" and mirrors
    // `ready` into Vue refs that gate ad-button visibility, so a
    // "watch ad" prompt never appears unless an ad is actually loaded.
    // The optional `error` carries the SDK's localizedDescription on
    // load/display failures so it surfaces in the on-device debug
    // overlay (mobile builds have no JS console).
    //
    // Two literal-dict call sites instead of a `[String: Any]` payload —
    // Tauri's `trigger(_:data:)` expects a JSObject ([String: JSValue]),
    // which Swift can synthesize from a homogeneous-ish literal but not
    // from an `[String: Any]` runtime value.
    private func emitReadiness(_ event: String, ready: Bool, error: String? = nil) {
        if let error = error {
            self.trigger(event, data: ["ready": ready, "error": error])
        } else {
            self.trigger(event, data: ["ready": ready])
        }
    }

    private func boolText(_ v: Bool) -> String { v ? "true" : "false" }

    @objc public func initialize(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(InitArgs.self)

        guard !args.appKey.isEmpty else {
            invoke.reject("levelplay: appKey is required")
            return
        }

        // ── COPPA / Families Policy metadata — MUST come before init ──
        LevelPlay.setMetaDataWithKey("is_child_directed", value: boolText(args.isChildDirected))
        LevelPlay.setMetaDataWithKey("AdMob_TFCD", value: boolText(args.admobTfcd))
        LevelPlay.setMetaDataWithKey("AdMob_TFUA", value: boolText(args.admobTfua))
        LevelPlay.setMetaDataWithKey("is_deviceid_optout", value: boolText(args.deviceIdOptOut))
        LevelPlay.setMetaDataWithKey("Meta_Mixed_Audience", value: boolText(args.metaMixedAudience))
        if args.enableTestSuite {
            LevelPlay.setMetaDataWithKey("is_test_suite", value: "enable")
        }

        let initRequest = LPMInitRequestBuilder(appKey: args.appKey).build()

        LevelPlay.initWith(initRequest) { [weak self] _, error in
            guard let self = self else { return }

            if let error = error {
                invoke.reject("levelplay init failed: \(error.localizedDescription)")
                return
            }

            self.initialized = true

            // Create per-format ad objects only when the caller supplied
            // a matching unit ID. Missing IDs leave the object nil and
            // surface as "not-configured" on show.
            if !args.rewardedAdUnitId.isEmpty {
                let rw = LPMRewardedAd(adUnitId: args.rewardedAdUnitId)
                rw.setDelegate(self)
                self.rewardedAd = rw
                self.rewardedAdUnitId = args.rewardedAdUnitId
                rw.loadAd()
            }

            if !args.interstitialAdUnitId.isEmpty {
                let itl = LPMInterstitialAd(adUnitId: args.interstitialAdUnitId)
                itl.setDelegate(self)
                self.interstitialAd = itl
                self.interstitialAdUnitId = args.interstitialAdUnitId
                itl.loadAd()
            }

            invoke.resolve(["initialized": true])
        }
    }

    @objc public func showRewarded(_ invoke: Invoke) throws {
        guard initialized else {
            invoke.reject("levelplay: not initialized")
            return
        }
        guard let rw = rewardedAd else {
            invoke.resolve(["shown": false, "rewarded": false, "error": "not-configured"])
            return
        }
        guard rw.isAdReady() else {
            rw.loadAd()
            invoke.resolve(["shown": false, "rewarded": false, "error": "no-fill"])
            return
        }
        guard pendingRewarded == nil else {
            invoke.resolve(["shown": false, "rewarded": false, "error": "already-showing"])
            return
        }
        pendingRewarded = invoke
        lastRewardGranted = false
        DispatchQueue.main.async { [weak self] in
            guard let vc = self?.topViewController else {
                let pending = self?.pendingRewarded
                self?.pendingRewarded = nil
                pending?.resolve([
                    "shown": false, "rewarded": false, "error": "no-view-controller",
                ])
                return
            }
            rw.showAd(viewController: vc, placementName: nil)
        }
    }

    // Opens LevelPlay's Test Suite overlay over the running app. Requires
    // `is_test_suite=enable` metadata to have been set BEFORE the SDK was
    // initialized (the JS provider does this when `enableTestSuite=true`
    // is passed to `init`). The Test Suite serves test ads regardless of
    // account-approval / store-publication status, which is the diagnostic
    // we need pre-launch.
    @objc public func launchTestSuite(_ invoke: Invoke) throws {
        guard initialized else {
            invoke.resolve(["launched": false, "error": "levelplay: not initialized"])
            return
        }
        DispatchQueue.main.async { [weak self] in
            guard let vc = self?.topViewController else {
                invoke.resolve(["launched": false, "error": "no-view-controller"])
                return
            }
            LevelPlay.launchTestSuite(vc)
            invoke.resolve(["launched": true])
        }
    }

    @objc public func showInterstitial(_ invoke: Invoke) throws {
        guard initialized else {
            invoke.reject("levelplay: not initialized")
            return
        }
        guard let itl = interstitialAd else {
            invoke.resolve(["shown": false, "error": "not-configured"])
            return
        }
        guard itl.isAdReady() else {
            itl.loadAd()
            invoke.resolve(["shown": false, "error": "not-ready"])
            return
        }
        guard pendingInterstitial == nil else {
            invoke.resolve(["shown": false, "error": "already-showing"])
            return
        }
        pendingInterstitial = invoke
        DispatchQueue.main.async { [weak self] in
            guard let vc = self?.topViewController else {
                let pending = self?.pendingInterstitial
                self?.pendingInterstitial = nil
                pending?.resolve([
                    "shown": false, "error": "no-view-controller",
                ])
                return
            }
            itl.showAd(viewController: vc, placementName: nil)
        }
    }

    // ── LPMRewardedAdDelegate ────────────────────────────────────────
    func didLoadAd(with adInfo: LPMAdInfo) {
        emitReadiness("rewarded-state", ready: true)
    }

    func didFailToLoadAd(withAdUnitId adUnitId: String, error: Error) {
        // Same selector signature is shared by both rewarded and
        // interstitial delegates in 9.x. We can't distinguish here
        // without unit-id matching, so emit both — JS only acts on the
        // ref that's actually backing a visible button.
        let errorMsg = error.localizedDescription
        if rewardedAdUnitId == adUnitId {
            emitReadiness("rewarded-state", ready: false, error: errorMsg)
        }
        if interstitialAdUnitId == adUnitId {
            emitReadiness("interstitial-state", ready: false, error: errorMsg)
        }
    }

    func didChangeAdInfo(_ adInfo: LPMAdInfo) {}
    func didDisplayAd(with adInfo: LPMAdInfo) {}

    func didFailToDisplayAd(with adInfo: LPMAdInfo, error: Error) {
        // Either format may emit this — dispatch by which pending
        // invoke is set. Only one show can be in flight per format at
        // a time (we reject overlap with "already-showing"), so a
        // non-nil pending slot identifies the active format.
        if let pending = pendingRewarded {
            pendingRewarded = nil
            lastRewardGranted = false
            pending.resolve([
                "shown": false, "rewarded": false,
                "error": error.localizedDescription,
            ])
            emitReadiness("rewarded-state", ready: false)
            rewardedAd?.loadAd()
        } else if let pending = pendingInterstitial {
            pendingInterstitial = nil
            pending.resolve([
                "shown": false,
                "error": error.localizedDescription,
            ])
            emitReadiness("interstitial-state", ready: false)
            interstitialAd?.loadAd()
        }
    }

    func didClickAd(with adInfo: LPMAdInfo) {}

    func didCloseAd(with adInfo: LPMAdInfo) {
        if let pending = pendingRewarded {
            pendingRewarded = nil
            let granted = lastRewardGranted
            lastRewardGranted = false
            pending.resolve(["shown": true, "rewarded": granted])
            emitReadiness("rewarded-state", ready: false)
            rewardedAd?.loadAd()
        } else if let pending = pendingInterstitial {
            pendingInterstitial = nil
            pending.resolve(["shown": true])
            emitReadiness("interstitial-state", ready: false)
            interstitialAd?.loadAd()
        }
    }

    func didRewardAd(with adInfo: LPMAdInfo, reward: LPMReward) {
        // Fired BEFORE didCloseAd. Just record; the resolve happens on
        // close so the caller knows playback ended (not just that the
        // reward was credited).
        lastRewardGranted = true
    }

    // ── LPMInterstitialAdDelegate (overlapping methods inherited above) ──
    // The interstitial delegate's didLoadAd / didFailToLoadAd /
    // didFailToDisplayAd / didCloseAd selectors share signatures with
    // the rewarded delegate, so the implementations above cover both.
}

@_cdecl("init_plugin_levelplay")
func initPlugin() -> Plugin {
    return LevelPlayPlugin()
}

File: src-tauri/plugins/tauri-plugin-levelplay/android/src/main/java/com/plugin/levelplay/LevelPlayPlugin.kt
package com.plugin.levelplay

import android.app.Activity
import android.webkit.WebView
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import com.unity3d.mediation.LevelPlay
import com.unity3d.mediation.LevelPlayAdError
import com.unity3d.mediation.LevelPlayAdInfo
import com.unity3d.mediation.LevelPlayConfiguration
import com.unity3d.mediation.LevelPlayInitError
import com.unity3d.mediation.LevelPlayInitListener
import com.unity3d.mediation.LevelPlayInitRequest
import com.unity3d.mediation.interstitial.LevelPlayInterstitialAd
import com.unity3d.mediation.interstitial.LevelPlayInterstitialAdListener
import com.unity3d.mediation.rewarded.LevelPlayReward
import com.unity3d.mediation.rewarded.LevelPlayRewardedAd
import com.unity3d.mediation.rewarded.LevelPlayRewardedAdListener

// JS → native payload for the init command. Shape matches
// `models::InitRequest` on the Rust side; Tauri auto-deserializes.
//
// `rewardedAdUnitId` / `interstitialAdUnitId` come from the LevelPlay
// dashboard under Ads → Ad Units. When empty, the matching show method
// returns a clean "not-configured" response so the rest of the app
// treats the call as a no-op instead of crashing.
@InvokeArg
class InitArgs {
    var appKey: String = ""
    var rewardedAdUnitId: String = ""
    var interstitialAdUnitId: String = ""
    var isChildDirected: Boolean = true
    var admobTfcd: Boolean = true
    var admobTfua: Boolean = true
    var deviceIdOptOut: Boolean = true
    var metaMixedAudience: Boolean = false
    var enableTestSuite: Boolean = false
}

// LevelPlay 9.x integration for Tauri.
//
// Two ordering rules for COPPA/Families-Policy compliance:
//
//   1. Every `LevelPlay.setMetaData(...)` call MUST happen BEFORE
//      `LevelPlay.init(...)`. The init call propagates the flags to each
//      adapter's SDK during startup; setting metadata afterwards is a
//      silent no-op for adapters that have already initialised.
//
//   2. Ad-format listeners (`LevelPlayRewardedAd.setListener` etc.) are
//      attached BEFORE we call `loadAd()` so the first load attempt
//      never races past the listener registration.
//
// The 9.x SDK uses per-instance ad objects (LevelPlayRewardedAd,
// LevelPlayInterstitialAd) keyed by dashboard-issued ad unit IDs. Each
// show method stores the active `Invoke` in a pending slot and resolves
// it exactly once from either `onAdClosed` or `onAdDisplayFailed`, so a
// concurrent second show can't cross-resolve an earlier caller's
// Promise. Lifecycle (`IronSource.onResume/onPause`) was removed in 9.x
// — the SDK now wires up `LevelPlayActivityLifecycleProvider` internally
// and we no longer need activity hooks on the plugin side.
@TauriPlugin
class LevelPlayPlugin(private val activity: Activity) : Plugin(activity) {

    private var initialized = false

    private var rewardedAd: LevelPlayRewardedAd? = null
    private var interstitialAd: LevelPlayInterstitialAd? = null

    private var pendingRewarded: Invoke? = null
    private var pendingInterstitial: Invoke? = null
    // `onAdRewarded` can fire before or after `onAdClosed`. We capture
    // the reward flag here and resolve on close so the caller knows
    // playback actually ended (not just that credit was granted).
    private var lastRewardGranted = false

    override fun load(webView: WebView) {
        super.load(webView)
    }

    @Command
    fun initialize(invoke: Invoke) {
        val args = invoke.parseArgs(InitArgs::class.java)

        if (args.appKey.isEmpty()) {
            invoke.reject("levelplay: appKey is required")
            return
        }

        // ── COPPA / Families Policy metadata — MUST come before init ──
        // See class doc for the ordering rule. Values are strings the
        // SDK parses as booleans from the literal "true" / "false" text.
        LevelPlay.setMetaData("is_child_directed", args.isChildDirected.boolText())
        LevelPlay.setMetaData("AdMob_TFCD", args.admobTfcd.boolText())
        LevelPlay.setMetaData("AdMob_TFUA", args.admobTfua.boolText())
        LevelPlay.setMetaData("is_deviceid_optout", args.deviceIdOptOut.boolText())
        LevelPlay.setMetaData("Meta_Mixed_Audience", args.metaMixedAudience.boolText())
        if (args.enableTestSuite) {
            LevelPlay.setMetaData("is_test_suite", "enable")
        }

        val initRequest = LevelPlayInitRequest.Builder(args.appKey).build()

        LevelPlay.init(activity, initRequest, object : LevelPlayInitListener {
            override fun onInitSuccess(configuration: LevelPlayConfiguration) {
                initialized = true

                // Create per-format ad objects only when the caller
                // supplied a matching unit ID. Missing IDs leave the
                // object null and surface as "not-configured" on show.
                if (args.rewardedAdUnitId.isNotEmpty()) {
                    val rw = LevelPlayRewardedAd(args.rewardedAdUnitId)
                    rw.setListener(rewardedListener)
                    rewardedAd = rw
                    rw.loadAd()
                }

                if (args.interstitialAdUnitId.isNotEmpty()) {
                    val itl = LevelPlayInterstitialAd(args.interstitialAdUnitId)
                    itl.setListener(interstitialListener)
                    interstitialAd = itl
                    itl.loadAd()
                }

                if (args.enableTestSuite) {
                    // `validateIntegration` writes the adapter-health
                    // report to logcat (under tag IntegrationHelper).
                    // `launchTestSuite` opens the interactive overlay
                    // that lets you trigger a test ad from any
                    // configured network on demand — this is what
                    // actually puts a test ad on screen without waiting
                    // for real fill. Both are debug-only.
                    LevelPlay.validateIntegration(activity)
                    activity.runOnUiThread {
                        LevelPlay.launchTestSuite(activity)
                    }
                }

                val res = JSObject()
                res.put("initialized", true)
                invoke.resolve(res)
            }

            override fun onInitFailed(error: LevelPlayInitError) {
                invoke.reject("levelplay init failed: ${error.errorMessage}")
            }
        })
    }

    // Emit a per-format readiness event to the webview. The JS provider
    // listens for "rewarded-state" / "interstitial-state" and mirrors
    // `ready` into Vue refs that gate ad-button visibility, so a
    // "watch ad" prompt never appears unless an ad is actually loaded.
    private fun emitReadiness(event: String, ready: Boolean) {
        val obj = JSObject()
        obj.put("ready", ready)
        trigger(event, obj)
    }

    // ── Rewarded video listener ──────────────────────────────────────
    private val rewardedListener = object : LevelPlayRewardedAdListener {
        override fun onAdLoaded(adInfo: LevelPlayAdInfo) {
            emitReadiness("rewarded-state", true)
        }

        override fun onAdLoadFailed(error: LevelPlayAdError) {
            emitReadiness("rewarded-state", false)
        }

        override fun onAdDisplayed(adInfo: LevelPlayAdInfo) {}

        override fun onAdDisplayFailed(error: LevelPlayAdError, adInfo: LevelPlayAdInfo) {
            val pending = pendingRewarded
            pendingRewarded = null
            lastRewardGranted = false
            val obj = JSObject()
            obj.put("shown", false)
            obj.put("rewarded", false)
            obj.put("error", error.errorMessage ?: "show-failed")
            pending?.resolve(obj)
            // Ad consumed/failed → not ready until reload completes.
            emitReadiness("rewarded-state", false)
            rewardedAd?.loadAd()
        }

        override fun onAdRewarded(reward: LevelPlayReward, adInfo: LevelPlayAdInfo) {
            lastRewardGranted = true
        }

        override fun onAdClicked(adInfo: LevelPlayAdInfo) {}
        override fun onAdInfoChanged(adInfo: LevelPlayAdInfo) {}

        override fun onAdClosed(adInfo: LevelPlayAdInfo) {
            val pending = pendingRewarded
            pendingRewarded = null
            val granted = lastRewardGranted
            lastRewardGranted = false
            val obj = JSObject()
            obj.put("shown", true)
            obj.put("rewarded", granted)
            pending?.resolve(obj)
            // Ad consumed → not ready; reload queues the next onAdLoaded
            // which will flip ready=true again.
            emitReadiness("rewarded-state", false)
            rewardedAd?.loadAd()
        }
    }

    // ── Interstitial listener ────────────────────────────────────────
    private val interstitialListener = object : LevelPlayInterstitialAdListener {
        override fun onAdLoaded(adInfo: LevelPlayAdInfo) {
            emitReadiness("interstitial-state", true)
        }

        override fun onAdLoadFailed(error: LevelPlayAdError) {
            emitReadiness("interstitial-state", false)
        }

        override fun onAdDisplayed(adInfo: LevelPlayAdInfo) {}

        override fun onAdDisplayFailed(error: LevelPlayAdError, adInfo: LevelPlayAdInfo) {
            val pending = pendingInterstitial
            pendingInterstitial = null
            val obj = JSObject()
            obj.put("shown", false)
            obj.put("error", error.errorMessage ?: "show-failed")
            pending?.resolve(obj)
            emitReadiness("interstitial-state", false)
            interstitialAd?.loadAd()
        }

        override fun onAdClicked(adInfo: LevelPlayAdInfo) {}
        override fun onAdInfoChanged(adInfo: LevelPlayAdInfo) {}

        override fun onAdClosed(adInfo: LevelPlayAdInfo) {
            val pending = pendingInterstitial
            pendingInterstitial = null
            val obj = JSObject()
            obj.put("shown", true)
            pending?.resolve(obj)
            emitReadiness("interstitial-state", false)
            interstitialAd?.loadAd()
        }
    }

    @Command
    fun showRewarded(invoke: Invoke) {
        if (!initialized) {
            invoke.reject("levelplay: not initialized")
            return
        }
        val rw = rewardedAd
        if (rw == null) {
            val obj = JSObject()
            obj.put("shown", false)
            obj.put("rewarded", false)
            obj.put("error", "not-configured")
            invoke.resolve(obj)
            return
        }
        if (!rw.isAdReady) {
            rw.loadAd()
            val obj = JSObject()
            obj.put("shown", false)
            obj.put("rewarded", false)
            obj.put("error", "no-fill")
            invoke.resolve(obj)
            return
        }
        if (pendingRewarded != null) {
            val obj = JSObject()
            obj.put("shown", false)
            obj.put("rewarded", false)
            obj.put("error", "already-showing")
            invoke.resolve(obj)
            return
        }
        pendingRewarded = invoke
        lastRewardGranted = false
        activity.runOnUiThread { rw.showAd(activity) }
    }

    // Opens LevelPlay's Test Suite overlay over the running Activity.
    // Requires `is_test_suite=enable` to have been set BEFORE init (the JS
    // provider does this when `enableTestSuite=true` is passed to `init`).
    // Test Suite serves test ads regardless of account-approval / store-
    // publication state — the right diagnostic for a pre-launch build.
    @Command
    fun launchTestSuite(invoke: Invoke) {
        if (!initialized) {
            val obj = JSObject()
            obj.put("launched", false)
            obj.put("error", "levelplay: not initialized")
            invoke.resolve(obj)
            return
        }
        activity.runOnUiThread {
            LevelPlay.launchTestSuite(activity)
            val obj = JSObject()
            obj.put("launched", true)
            invoke.resolve(obj)
        }
    }

    @Command
    fun showInterstitial(invoke: Invoke) {
        if (!initialized) {
            invoke.reject("levelplay: not initialized")
            return
        }
        val itl = interstitialAd
        if (itl == null) {
            val obj = JSObject()
            obj.put("shown", false)
            obj.put("error", "not-configured")
            invoke.resolve(obj)
            return
        }
        if (!itl.isAdReady) {
            itl.loadAd()
            val obj = JSObject()
            obj.put("shown", false)
            obj.put("error", "not-ready")
            invoke.resolve(obj)
            return
        }
        if (pendingInterstitial != null) {
            val obj = JSObject()
            obj.put("shown", false)
            obj.put("error", "already-showing")
            invoke.resolve(obj)
            return
        }
        pendingInterstitial = invoke
        activity.runOnUiThread { itl.showAd(activity) }
    }
}

// Locale-safe boolean → "true"/"false" string for LevelPlay metadata.
// `Boolean.toString()` on JVM is locale-independent, but this keeps the
// intent explicit so nobody reaches for `String.format` in a future edit.
private fun Boolean.boolText(): String = if (this) "true" else "false"
